---
title: "Practicum 17: Gene Expression Microarrays"
author: "Data Science for Biomedical Informatics (BMIN503/EPID600)"
output:
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---

```{r global options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
``` 

***
New packages to install for this practicum:
```{r, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("GEOquery")
biocLite("oligo")
biocLite("limma")
biocLite("hgu133plus2.db")
biocLite("pd.hg.u133.plus.2")
biocLite("viridis")
install.packages("gplots")
```
***

We will use GEO dataset [GSE8823](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE8823) to learn some basics about expression microarray analysis. Note that you could use [GEO2R](http://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE8823) to perform some of the analysis via a web interface that runs R on the back-end. If you are considering a more detailed gene expression microarray analysis, take a look at [raved](https://github.com/HimesGroup/raved).

Follow the study link and notice where some of the information we will use comes from:

* Study design
* Platform used
* Paper associated with the GEO entry


Let's load the necessary libraries and get the raw files to begin.
```{r eval=FALSE}
library(GEOquery)
getGEOSuppFiles("GSE8823")
untar("./GSE8823/GSE8823_RAW.tar", exdir="./GSE8823/data")
```

```{r eval=TRUE, message=FALSE, warning=FALSE}
library(oligo)
library(limma)
library(viridis)

celFiles <- list.celfiles("./GSE8823/data", full.names=TRUE, listGzipped=TRUE)
raw.data <- read.celfiles(celFiles)
pData(raw.data)

pData(raw.data)$status <- c(rep("nonsmoker", 11), rep("smoker",13))

pData(raw.data)

boxplot(raw.data, col="red", main="Raw Probe Intensities")

GSE8823.rma <- rma(raw.data)

boxplot(GSE8823.rma, col="blue", main="RMA Expression Values")
```

Looking at the raw expression, notice that the intensities for each sample vary widely. This does not reflect the biological variation we are interested in, so to adjust the scales, we can perform RMA. A more comprehensive QC report could be obtained using `arrayQualityMetrics` or [raved](https://github.com/HimesGroup/raved). Selecting which samples to exclude from analysis is done after looking at quantitative and objective data, but is a bit of an art. First, we will drop a sample based on its raw boxplot. Then, to perform differential expression analysis, we will use `limma`. First, we create a design model matrix and specify which groups we would like to contrast.
```{r eval=TRUE}
GSE8823.qc <- raw.data[, !sampleNames(raw.data) %in% "GSM219255.CEL.gz"]
GSE8823.qc.rma <- rma(GSE8823.qc)
design <- model.matrix(~ -1+factor(GSE8823.qc.rma$status))
colnames(design) <- levels(factor(GSE8823.qc.rma$status))
design

#Fit a linear model, according to limma package implementation, to expression data using design matrix model
fit <- lmFit(GSE8823.qc.rma, design)
#Create contrast groups of interest
GSE8823.contrast <- makeContrasts(smoking = smoker-nonsmoker, 
                                  levels = design)
                                 
#Get the contrasts for samples of interest
fit2 <- contrasts.fit(fit, GSE8823.contrast)
#Adjust fit coefficients using an empirical Bayes moderation of standard errors
fit2 <- eBayes(fit2)

#Results for each hypothesis test can be extracted using
smoking_results <- topTable(fit2, coef="smoking", adjust="BH", num=200)
head(smoking_results)
```

The row names are probes, but it would be nice to have official gene symbols. If you have a long list of results, this look-up can be time consuming.
```{r eval=TRUE, message=FALSE}
library(hgu133plus2.db)
smoking_results$ID = row.names(smoking_results)
smoking_results$SYMBOL <- lapply(smoking_results$ID, function(x) mget(x, env=hgu133plus2SYMBOL, ifnotfound=NA)[[1]])
head(smoking_results)
```

To get ideas of how to present data, let's make a volcano plot, a boxplot for one probe and a heatmap of the top results. First, we will generate a volcano plot using ggplot2.

```{r eval=TRUE}
library(ggplot2)
# Output all differential expression results
full_res <- topTable(fit2, coef="smoking" , adjust="BH", num=Inf)
# Assign (in)significant genes
full_res$sig <- rep("insignificant", nrow(full_res))
full_res$sig[which(full_res$adj.P.Val<0.05)] <- "significant"
ggplot(full_res, aes(x=logFC, y=-log10(adj.P.Val), color=sig)) + geom_point() +
  theme_bw() +
  ggtitle("Volcano plot") +
  xlab("logFC")+
  ylab("-log10(q-value)") +
  scale_color_manual(values=c("black","red")) +
  theme(legend.position="none")
# A volcano plot could also be generated using a built-in limma function:
# volcanoplot(fit2, coef="smoking", highlight=10)
```

Next, let's create a boxplot for one probe
```{r eval=TRUE}
# make a data frame for gene expression of the probe
df_boxplot=data.frame(
  expression=exprs(GSE8823.qc.rma)["1556180_at", ],
  status=GSE8823.qc.rma$status)

ggplot(df_boxplot, aes(x=status,y=expression)) +
  geom_boxplot(outlier.colour=NA, color="grey18", fill="lightgreen") +
  stat_boxplot(geom='errorbar', color="grey18") +
  geom_jitter(size=1, position=position_jitter(width=0.3)) +
  ggtitle("Boxplot for probe 1556180_at") +
  xlab("")+
  ylab("RMA-intensity")+
  theme_bw() +
  theme(legend.position="none")
```

Finally, let's generate a heatmap for the top 200 probes using gplots
```{r eval=TRUE, fig.width=12, fig.height=10}
library(gplots)
top.eset <- GSE8823.qc.rma[row.names(exprs(GSE8823.qc.rma)) %in% row.names(smoking_results)]
heatmap.2(exprs(top.eset), trace="none", keysize=1.5, key.title=NA)
#Add status to top of plot
status.colors <- unlist(lapply(GSE8823.qc.rma$status, function(x){if (x=="smoker") "navy" else "red"}))
heatmap.2(exprs(top.eset), col=viridis(256, option="B"),
          trace="none", keysize=1.5, key.title=NA,
          ColSideColors=status.colors)
legend("topright", legend=c("smoker", "nonsmoker"), fill=c("red", "navy")) 
```



