---
title: "Assignment 6"
author: "Data Science for Biomedical Informatics (BMIN503/EPID600)"
output:
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
***

#### Instructions
- Download the Rmd version of this file
- Complete the questions below in RStudio using the Rmd file as a template 
- Replace text in brackets with your answers. (There is no need to keep the brackets).
- Save the Rmd document as Assignment6_*YOUR LAST NAME*.Rmd
- Create an HTML version using knitr
- Turn in completed Assignment6_*YOUR LAST NAME*.html file in Canvas under Assignments -> Assignment 6
- Your assignment **must** be in html format or it will not be graded
- Grades will be assigned according to point scheme below and computed as a percentage of total possible points
- Lateness policy: If an urgent and unforseen circumstance comes up and you need an extension to turn in an assignment, please contact Blanca and/or Sherrie as soon as possible. Unless there is an arrangement in place, late assignments will be scored as follows: 
    - 25% of total score if 1 day late
    - 50% of total score if 2 days late
    - 75%  of total score if 3 days late
    - Assignment will not be graded if 4 or more days late
- DUE DATE: 11/27/18


#### Final Project - Methods and Results Draft
1. Using your local forked [Final Project Repo](https://github.com/HimesGroup/BMIN503_Final_Project), write a draft of the Methods and Results sections following the instructions provided in the Rmd file where you drafted the Introduction. You should be using an RStudio Project to pull/push changes as you work. Edit your own README.md file to describe what your project is about and what files are present in the repository.

> ## Methods
> The work of this project was split into two general parts, ETL and Model Building.

> ### Extract, Transform, Load (ETL)
> Firstly I utilized python extensively to download, format and parse my data into the full training dataset.

> My source was firstly the XML formatted records of each protein manually annotated and curated within Uniprot. These files contained the approved sequences and other data on each protein in uniprot. The other second data source is the amyloid database at amypro.net, this dataset has the amyloid sequences of every protein found to form amyloid protein.

> To process the protein sequences I utilized the biopython package to perform a number of calculations on each protein sequence and extract a number of protein-level features. To n-gram each sequence I used the nltk python package to process the protein sequences.

> There was some effort in extracting, using NLP, the comments and other features included in the XML other than the protein sequence. These features ended up being difficult to parse, contained little information and were ultimately excluded from the final dataset.

> Further work could be done in performing a gene enrichment pipeline on all of the proteins and including those data points in the final training data.

> ## Results
> The original setup of the problem and its answer resulted in a model that predicted whether entire proteins were likely to produce amyloid. The resulting model produced a 95% AUC-ROC score across a few different permutations of the data.

> This result is explained by the data I used to train it. I basically duplicated the protein level features by n-gramming, making the protein level features much more important than the n-gram level features. If a protein had 3000 n-grams, then the protein level features were duplicated 3000 times. Ultimately this result was unsatisfactory to me, so I re-formatted my problem.

> What I'm more interested in is if I can point at any given section of protein and predict whether it will form amyloid. This is the problem I am working on now (and hopefully will finish by my presentation). To achieve a model that does this I am adding many more n-gram level features to balance the data between protein-level features and ngram-level features. I am also changing the prediction output to give me an array of n-predictions. In my case the output could be [1, 0, 0, 1, 1] which would mean the first and last 2 amino acids are predicted as amyloid.

```{r include=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite(c(
  "GEOquery", "oligo", "limma", "hgu133plus2.db", 
  "pd.hg.u133.plus.2", "viridis", "gplots","oligoClasses","BiocParallel"
  )
)

```

#### Gene Expression
2. Analyse GEO dataset [GSE27973](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE27973) to identify genes that are differentially expressed in airway epithelial cells in response to cigarette smoke extract (CSE) exposure. The procedure you need to follow is similar to that followed in Practicum 16. Specifically,
    + Download the raw CEL files from GEO, create an expression set, and include a phenotype variable corresponding to `treatment` with four levels as can be inferred from the GEO submission.  (*2 points*)
    
    
```{r eval=TRUE}
library(limma)
library(GEOquery)

getGEOSuppFiles("GSE27973")
untar("./GSE27973/GSE27973_RAW.tar", exdir="./GSE27973/data")
```


```{r, eval=TRUE}
library(oligoClasses)
library(oligo)
library(viridis)

celFiles <- list.celfiles("./GSE27973/data", full.names=TRUE, listGzipped=TRUE)
raw.data <- read.celfiles(celFiles)
pData(raw.data)

pData(raw.data)$treatment <- c(rep("medium"), rep("RV16"), rep("CSE"), rep("RV16_and_CSE"))

pData(raw.data)
```
    
    
    + Look at a boxplot of expression values for each sample and see whether any seem to be outliers. (*2 points*)
    
```{r, eval=TRUE}
par(mar=c(9,5,1,1))
boxplot(raw.data, col="red", main="RAW Expression Values", las=2)
```
    
> It looks like the samples vary quite a bit, they all seem to have really high error bars too. In terms out 'outliers' I might point at GSM692115 for being high, GSM692118 for being low and GSM69219 for being high - its hard to say for certain though.
    
    + Perform RMA on the samples and create a new intensity boxplot. (*2 points*)
    
```{r, eval=TRUE}
GSE27973.rma <- rma(raw.data)

par(mar=c(9,5,1,1))
boxplot(GSE27973.rma, col="blue", main="RMA Expression Values", las=2)
```
    
> This looks a lot more uniform! Thanks RMA normalization ;)
    
    + Perform a differential expression analysis, and report how many probes are significantly differentially expressed between the *CSE* and *medium* samples based on a Benjamini-Hochberg corrected p-value threshold of 0.05. Note: it is OK to do an analysis without taking into account donor status. That is, compare sample expression based on treatment status only. (*3 points*)
    
```{r, eval=TRUE}
GSE27973.qc <- raw.data[, !sampleNames(raw.data) %in% "GSE27973.CEL.gz"]
GSE27973.qc.rma <- rma(GSE27973.qc)
design <- model.matrix(~ -1+factor(GSE27973.qc.rma$treatment))
colnames(design) <- levels(factor(GSE27973.qc.rma$treatment))
design

#Fit a linear model, according to limma package implementation, to expression data using design matrix model
fit <- lmFit(GSE27973.qc.rma, design)
#Create contrast groups of interest
GSE27973.contrast <- makeContrasts(treatment = CSE-medium, 
                                  levels = design)
                                 
#Get the contrasts for samples of interest
fit2 <- contrasts.fit(fit, GSE27973.contrast)
#Adjust fit coefficients using an empirical Bayes moderation of standard errors
fit2 <- eBayes(fit2)

#Results for each hypothesis test can be extracted using
treatment_results <- topTable(fit2, coef="treatment", adjust="BH", p.value = 0.05)
head(treatment_results)

nrow(treatment_results)
```
    
> There are only 10 probes that show a statistically different expression level between samples treated with CSE-only versus samples treated with medium-only.
    
    + What is the top differentially expressed gene? Create a boxplot of its normalized intensities across **all** treatment groups. How many probes corresponding to this gene are among the top hits? (*3 points*)
    
```{r, eval=TRUE}
# Get gene names
library(hgu133plus2.db)
library(ggplot2)
library(dplyr)

# Output all differential expression results
full_res <- topTable(fit2, coef="treatment" , adjust="BH", num=Inf)
# Assign (in)significant genes
full_res$sig <- rep("insignificant", nrow(full_res))
full_res$sig[which(full_res$adj.P.Val<0.05)] <- "significant"
full_res.significant <- full_res[full_res$sig == "significant",]
full_res.significant$ID = row.names(full_res.significant)
full_res.significant$SYMBOL <- lapply(full_res.significant$ID, function(x) mget(x, env=hgu133plus2SYMBOL, ifnotfound=NA)[[1]])

full_res.significant <- full_res.significant[order(full_res.significant$P.Value),]

top_probe <- rownames(full_res.significant)[1]
top_gene <- full_res.significant$SYMBOL[[1]]
print(paste0("The gene with the highest likelihood of its differential expression not being accounted for by chance (lowest p-val): ",top_gene))

num_probes_corr_to_top_gene <- nrow(full_res.significant[full_res.significant$SYMBOL==top_gene,])
print(paste0("The number of probes corresponding to ",top_gene,": ",num_probes_corr_to_top_gene))

df_boxplot=data.frame(
  expression=exprs(GSE27973.qc.rma)[top_probe, ],
  treatment=GSE27973.qc.rma$treatment)

ggplot(df_boxplot, aes(x=treatment,y=expression)) +
  geom_boxplot(outlier.colour=NA, color="grey18", fill="lightgreen") +
  stat_boxplot(geom='errorbar', color="grey18") +
  geom_jitter(size=1, position=position_jitter(width=0.3)) +
  ggtitle(paste("Boxplot for probe ", top_probe)) +
  xlab("")+
  ylab("RMA-intensity")+
  theme_bw() +
  theme(legend.position="none")
```

> The top gene is "NQO1" and it has 3 probes associated in the most significant results. The above probe is the most differentially expressed one in the dataset between all treatment groups.
    
    + Create a heatmap using the significantly differentially expressed genes and samples from all conditions. How do the treatment groups cluster? What does this indicate (a qualitative answer is sufficient)? (*3 points*)
    
```{r eval=TRUE, fig.width=12, fig.height=10}
library(gplots)
top.eset <- GSE27973.qc.rma[row.names(exprs(GSE27973.qc.rma)) %in% row.names(full_res.significant)]
heatmap.2(exprs(top.eset), trace="none", keysize=1.5, key.title=NA, margins = c(11,8))
#Add status to top of plot
status.colors <- unlist(lapply(GSE27973.qc.rma$treatment, function(x){if (x=="CSE") "navy" else if (x=="medium") "red" else if (x=="RV16") "orange" else "yellow"}))
par(mar=c(7,4,4,2)+0.1) 
heatmap.2(exprs(top.eset), col=viridis(256, option="B"),
          trace="none", keysize=1.5, key.title=NA,
          ColSideColors=status.colors, margins = c(11,8))
legend("topright", legend=c("CSE", "medium", "RV16", "RV16_and_CSE"), fill=c("navy", "red", "orange", "yellow")) 
```

> It appears that the medium and RV16 tend to express similarly, for just about all markers. Similarly, the CSE and RV16_and_CSE treatment groups tend to group together on all expression markers. Some genes are highly expressed in one group or the other, or vice versa, with all groups appearing to similarly express 204050_s_at and the three "X_at" genes near the middle highlighted in yellow. Similarly, the final expression marker appears to be consistently expressed at a low rate across all groups.

> The starkest difference between the groups appears to be at "205749_at" as well as with "201468_s_at". The other markers appear to be expressed at high or low rates, but at different magnitudes.